### PR TITLE
[wgsl-in] Handle modf and frexp

### LINKED ISSUE #594: Figure out modf story
See https://github.com/gpuweb/gpuweb/issues/1480
Currently, our backends are not producing good code for this.

### LINKED ISSUE #1161: Change frexp and modf to work with structures
See https://github.com/gpuweb/gpuweb/pull/1973
I think this change needs to be done at IR level.

### PR DESCRIPTION
**Pull request by fornwall**
Fixes #1161 (also closes #594 & closes #1441).

A bit rough, will check it more thoroughly, but opening it for feedback.

Moves `__atomic_compare_exchange_result<{kind:?},{width}>` to the new `special_types` as it probably fits in there (and gets rid of a TODO comment).

### CODE REVIEW

**Comment by fornwall** [comment]
The build failure is #2447.

File: src/front/type_gen.rs [hunk 1]
```diff
@@ -311,4 +262,203 @@ impl crate::Module {
...
+                        },
+                        Span::UNDEFINED,
+                    );
+                    let vec_int_ty = self.types.insert(
+                        crate::Type {
+                            name: None,
+                            inner: crate::TypeInner::Vector {
+                                size,
+                                kind: crate::ScalarKind::Sint,
+                                width: 4,
```
Comment (teoxoy) [inline src/front/type_gen.rs:419]:
```suggestion
                                width,
```

File: src/front/type_gen.rs [hunk 2]
```diff
@@ -311,4 +262,203 @@ impl crate::Module {
...
+                            },
+                        },
+                        Span::UNDEFINED,
+                    );
+                    (vec_float_ty, vec_int_ty, size as u32 * width as u32)
+                } else {
+                    (float_ty, int_ty, 4)
```
Comment (teoxoy) [inline src/front/type_gen.rs:426]:
```suggestion
                    (float_ty, int_ty, width as u32)
```

File: src/front/type_gen.rs [hunk 3]
```diff
@@ -311,4 +262,203 @@ impl crate::Module {
...
+                                size,
+                                kind: crate::ScalarKind::Float,
+                                width,
+                            },
+                        },
+                        Span::UNDEFINED,
+                    );
+                    (vec_ty, size as u32 * width as u32)
+                } else {
+                    (float_ty, 4)
```
Comment (teoxoy) [inline src/front/type_gen.rs:348]:
```suggestion
                    (float_ty, width as u32)
```

File: src/back/wgsl/writer.rs [hunk 4]
```diff
@@ -96,6 +96,13 @@ impl<W: Write> Writer<W> {
         self.ep_results.clear();
     }
 
+    fn is_builtin_wgsl_struct(&self, ty_struct: &crate::Type) -> bool {
+        ty_struct
+            .name
+            .as_ref()
+            .map_or(false, |name| name.starts_with("__"))
```
Comment (teoxoy) [inline src/back/wgsl/writer.rs:103]: Can we instead check it's inside `predeclared_types`? It should be more robust.

File: src/back/hlsl/help.rs [hunk 5]
```diff
@@ -781,6 +781,59 @@ impl<'a, W: Write> super::Writer<'a, W> {
...
+
+                    let (defined_func_name, called_func_name, second_field_name, sign_multiplier) =
+                        if matches!(type_key, &crate::PredeclaredType::ModfResult { .. }) {
+                            (super::writer::MODF_FUNCTION, "modf", "whole", "")
+                        } else {
+                            (
+                                super::writer::FREXP_FUNCTION,
+                                "frexp",
+                                "exp_",
+                                "sign(arg) * ",
```
Comment (teoxoy) [inline src/back/hlsl/help.rs:812]: Doesn't seem to be mentioned in the docs that HLSL's `frexp`'s result will always be positive.

Was this found by testing it?

**Review by teoxoy — COMMENTED** [review]
Really great stuff!

File: src/back/wgsl/writer.rs (hunk 4 above)
Comment (fornwall) [reply src/back/wgsl/writer.rs]: Good idea, done in e8fe048dd3c52064bdcb9db7d8bc39dbeb60f284

File: src/back/hlsl/help.rs (hunk 5 above)
Comment (fornwall) [reply src/back/hlsl/help.rs:812]: Yes - this was found by testing.

**Review by teoxoy — APPROVED** [review]

**Pull request merged**